src/numerics/include/metaphysicl/dualnumber.h

```diff
@@ -274,6 +324,125 @@ D gradient(const DualNumber<T, D>& a)
   return a.derivatives();
 }
 
+//
+// DualNumber helper function definitions
+//
+
+template <typename T, typename TBase, size_t N, typename R, class... PtrArgs, class... ParamArgs>
+DualNumber<R, NumberArray<N, R>>
+return_dn(const R & (TBase::*fn)(PtrArgs...) const,
+          const DualNumber<T, NumberArray<N, T>> & calling_dn,
+          ParamArgs &&... args)
+{
+  NumberArray<N, R> deriv;
+  for (decltype(N) di = 0; di < N; ++di)
+    deriv[di] = (calling_dn.derivatives()[di].*fn)(std::forward<ParamArgs>(args)...);
+  return {(calling_dn.value().*fn)(std::forward<ParamArgs>(args)...), deriv};
+}
+
+template <typename T, typename TBase, size_t N, typename R, class... PtrArgs, class... ParamArgs>
+DualNumber<R, NumberArray<N, R>>
+return_dn(R (TBase::*fn)(PtrArgs...) const,
+          const DualNumber<T, NumberArray<N, T>> & calling_dn,
+          ParamArgs &&... args)
+{
+  NumberArray<N, R> deriv;
+  for (decltype(N) di = 0; di < N; ++di)
+    deriv[di] = (calling_dn.derivatives()[di].*fn)(std::forward<ParamArgs>(args)...);
+  return {(calling_dn.value().*fn)(std::forward<ParamArgs>(args)...), deriv};
+}
+
+template <typename T, typename TBase, size_t N, typename R, class... PtrArgs, class... ParamArgs>
+DualNumberSurrogate<R, N> &
+return_dns(R & (TBase::*)(PtrArgs...),
+           DualNumber<T, NumberArray<N, T>> & calling_dn,
+           ParamArgs &&... args)
+{
+  std::tuple<PtrArgs...> key(std::forward<ParamArgs>(args)...);
+  typename std::map<decltype(key), DualNumberSurrogate<R, N>>::iterator it =
+      calling_dn.dual_number_surrogates.find(key);
```

Comment:
Either github or my browser is failing me; when I search for dual_number_surrogates these four lines are the only instances I find...

Reply:
You are not crazy. This is also pertinent to a more general area of concern, which was the initial reason I created a `DualNumberBase`: we're going to have users who are going to call methods like `tensor_construct.tr()` on a property that used to be a `RealTensorValue` and is now a `DualNumber<RealTensorValue>` and expect it to work. This was the reason for the creation of all the forwarding methods and free functions here. So what I need are `DualNumber` specializations that have tensor/vector specializations. It seems really foolish to have to duplicate all the general `DualNumber` template member declaration/definitions for the specializations. Do you have a suggestion for the best way to handle this? I'm still growing as a C++ programmer...is this a place for traits techniques?

Going to your particular question...in libmesh/MOOSE I have these `DualNumber` vector/tensor specializations that have a map member that stores `DualNumberSurrogates` with keys corresponding to a tuple of the vector/tensor indices. Example here:
```c++
#ifndef DUALNUMBERRANKTWOTENSOR_H
#define DUALNUMBERRANKTWOTENSOR_H

#include "metaphysicl/dualnumber.h"
#include "RankTwoTensor.h"

template <typename T, std::size_t N>
class DualNumber<T, N, typename std::enable_if<std::is_same<T, RankTwoTensor>::value>::type>
  : public DualNumberBase<T, NumberArray<N, T>>
{
public:
  using DualNumberBase<T, NumberArray<N, T>>::DualNumberBase;

  DualNumber<T, N> operator-() const { return DualNumber<T, N>(-this->_val, -this->_deriv); }
  DualNumber<T, N> operator!() const { return DualNumber<T, N>(!this->_val, !this->_deriv); }

  template <typename T2>
  DualNumber(const DualNumber<TypeVector<T2>, N> & row1,
             const DualNumber<TypeVector<T2>, N> & row2,
             const DualNumber<TypeVector<T2>, N> & row3)
    : DualNumberBase<T, NumberArray<N, T>>()
  {
    this->value() = RankTwoTensor(row1.value(), row2.value(), row3.value());
    for (decltype(N) i = 0; i < N; ++i)
      this->derivatives()[i] =
          RankTwoTensor(row1.derivatives()[i], row2.derivatives()[i], row3.derivatives()[i]);
  }
  template <typename T2>
  DualNumber(const DualNumber<VectorValue<T2>, N> & row1,
             const DualNumber<VectorValue<T2>, N> & row2,
             const DualNumber<VectorValue<T2>, N> & row3)
    : DualNumberBase<T, NumberArray<N, T>>()
  {
    this->value() = RankTwoTensor(row1.value(), row2.value(), row3.value());
    for (decltype(N) i = 0; i < N; ++i)
      this->derivatives()[i] =
          RankTwoTensor(row1.derivatives()[i], row2.derivatives()[i], row3.derivatives()[i]);
  }

  ConstReturnDecl(operator());
  NonConstReturnDecl(operator());
  NonConstVoidDecl(zero);
  ConstReturnDecl(row);
  ConstReturnDecl(tr);
  ConstReturnDecl(transpose);
  NonConstVoidDecl(rotate);
  ConstReturnDecl(positiveProjectionEigenDecomposition);
  ConstVoidDecl(symmetricEigenvaluesEigenvectors);
  ConstReturnDecl(deviatoric);

  std::map<typename T::index_type, DualNumberSurrogate<typename T::value_type, N>>
      dual_number_surrogates;
};

ConstReturnDef(operator(), std::is_same<T COMMA RankTwoTensor>::value);
NonConstReturnDef(operator(), std::is_same<T COMMA RankTwoTensor>::value);
NonConstVoidDef(zero, std::is_same<T COMMA RankTwoTensor>::value);
ConstReturnDef(row, std::is_same<T COMMA RankTwoTensor>::value);
ConstReturnDef(tr, std::is_same<T COMMA RankTwoTensor>::value);
ConstReturnDef(transpose, std::is_same<T COMMA RankTwoTensor>::value);
NonConstVoidDef(rotate, std::is_same<T COMMA RankTwoTensor>::value);
ConstReturnDef(positiveProjectionEigenDecomposition, std::is_same<T COMMA RankTwoTensor>::value);
ConstVoidDef(symmetricEigenvaluesEigenvectors, std::is_same<T COMMA RankTwoTensor>::value);
ConstReturnDef(deviatoric, std::is_same<T COMMA RankTwoTensor>::value);

#endif
```
(I have no idea why I made `dual_number_surrogates` public. That needs to be changed...)

Reply:
> My solution to that problem in the past was to always use DualNumber as an "inner" class within "outer" classes like tensors. That relies on the outer class methods being wholly available as generics, however, which requirement is usually inconvenient and occasionally impossible. But that's one explanation for why DualNumber only properly wraps other-things-that-duck-type-as-numbers.

So using the `DualNumber` as the inner may still be what we go to...but our thought (pulling in @friedmud here) was such:

We have thousands of `Kernels` already out in the MOOSE world and they all (the `Kernels`) expect when they ask for property `tensor_foo` that they're getting a say `RealTensorValue`. These `Kernels` already have (hopefully accurate) hand-coded Jacobians. Now I'm an application developer and I come along and I want to make use of a lot of these existing `Kernels`. But I am also going to create a a few new `Kernels` that are also going to take `tensor_foo` **and** I don't want to hand-code any Jacobians. I'm going to create a a new `Material` that calculates the property `tensor_foo` as a tensor dual number. I desire to only calculate `tensor_foo` once, and then I want to pass `tensor_foo` as a dual number to my new `Kernels` but as a raw tensor to my old `Kernels`. If `DualNumber` is the outer class, it's "easy" to return `tensor_foo.value()` (or `raw_value(tensor_foo)`!) when we want the raw tensor, but if `DualNumber` is the inner class, then what do we do? I suppose if the `Kernels` are not operating on `tensor_foo` then I could just do something like this from the `Material`:

```c++
    return TensorValue(tensor_foo(0, 0).value(), tensor_foo(0, 1).value(), ...)
```

Ugh, my head is spinning...I may not have been clear. And I'm not convinced myself anymore we need `DualNumber` as the outer class. Maybe @friedmud can motivate it more.